File: Silver/2016/February/03_milk_pails/MilkPails.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class MilkPails {
    static int x, y, k, m;
    static HashSet<List<Integer>> possibleCombination = new HashSet<>();

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("pails.in"));
        PrintWriter pw = new PrintWriter("pails.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        x = Integer.parseInt(st.nextToken());
        y = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        possibleCombination.add(Arrays.asList(0, 0));
        // 从当前可能的状态集合出发操作一次得到下一个可能的状态集合，重复 k 次
        for (int i = 0; i < k; i++) {
            HashSet<List<Integer>> temp = new HashSet<>();
            for (List<Integer> cur : possibleCombination) {
                List<List<Integer>> coms = nextCombination(cur);
                temp.addAll(coms);
            }
            possibleCombination = temp;
        }
        int ans = m;
        for (List<Integer> cur : possibleCombination) {
            ans = Math.min(ans, Math.abs(cur.get(0) + cur.get(1) - m));
        }
        pw.println(ans);
        pw.close();
    }

    /**
     * 从当前状态得到下一个状态 只有6种可能
     * 1. 倒空 x
     * 2. 倒空 y
     * 3. 倒满 x
     * 4. 倒满 y
     * 5. x 倒入 y    分两种情况：x + y > y 时候 y 倒满，否则 x 倒空
     * 6. y 倒入 x    分两种情况：x + y > x 时候 x 倒满，否则 y 倒空
     */
    static List<List<Integer>> nextCombination(List<Integer> cur) {
        List<List<Integer>> coms = new ArrayList<>();
        List<Integer> com1 = Arrays.asList(0, cur.get(1));
        List<Integer> com2 = Arrays.asList(cur.get(0), 0);
        List<Integer> com3 = Arrays.asList(x, cur.get(1));
        List<Integer> com4 = Arrays.asList(cur.get(0), y);
        List<Integer> com5;
        if (cur.get(0) + cur.get(1) > x) {
            com5 = Arrays.asList(x, cur.get(1) - (x - cur.get(0)));
        } else {
            com5 = Arrays.asList(cur.get(0) + cur.get(1), 0);
        }
        List<Integer> com6;
        if (cur.get(0) + cur.get(1) > y) {
            com6 = Arrays.asList(cur.get(0) - (y - cur.get(1)), y);
        } else {
            com6 = Arrays.asList(0, cur.get(0) + cur.get(1));
        }
        coms.add(com1);
        coms.add(com2);
        coms.add(com3);
        coms.add(com4);
        coms.add(com5);
        coms.add(com6);
        return coms;
    }
}

File: Silver/2016/February/03_milk_pails/milk_pails.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;
struct pair_hash {
    size_t operator()(const pair<int, int> &pair) const {
        size_t hash1 = hash<int>{}(pair.first);
        size_t hash2 = hash<int>{}(pair.second);
        return hash1 * 1000 + hash2;
    }
};
int x, y, k, m;
unordered_set<pair<int, int>, pair_hash> possible_combination;

// 从当前状态得到下一个状态的函数
vector<pair<int, int>> next_combination(const pair<int, int>& cur) {
    vector<pair<int, int>> coms;

    // 1. 倒空 x
    coms.push_back({0, cur.second});
    // 2. 倒空 y
    coms.push_back({cur.first, 0});
    // 3. 倒满 x
    coms.push_back({x, cur.second});
    // 4. 倒满 y
    coms.push_back({cur.first, y});
    // 5. x 倒入 y
    if (cur.first + cur.second > x) {
        coms.push_back({x, cur.second - (x - cur.first)});
    } else {
        coms.push_back({cur.first + cur.second, 0});
    }
    // 6. y 倒入 x
    if (cur.first + cur.second > y) {
        coms.push_back({cur.first - (y - cur.second), y});
    } else {
        coms.push_back({0, cur.first + cur.second});
    }

    return coms;
}

int main() {
    ifstream fin("pails.in");
    ofstream fout("pails.out");

    fin >> x >> y >> k >> m;
    possible_combination.insert({0, 0});

    // 从当前可能的状态集合出发操作一次得到下一个可能的状态集合，重复 k 次
    for (int i = 0; i < k; i++) {
        unordered_set<pair<int, int>, pair_hash> temp;
        for (const pair<int, int>& cur : possible_combination) {
            vector<pair<int, int>> coms = next_combination(cur);
            temp.insert(coms.begin(), coms.end());
        }
        possible_combination = temp;
    }

    int ans = m;

    // 遍历所有可能的状态，计算最小差值
    for (const pair<int, int>& cur : possible_combination) {
        ans = min(ans, abs(cur.first + cur.second - m));
    }

    fout << ans << endl;
    fout.close();

    return 0;
}


